wrapper/src/main/java/software/amazon/jdbc/plugin/bluegreen/BlueGreenStatusMonitor.java

```diff
@@ -0,0 +1,640 @@
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.plugin.bluegreen;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLSyntaxErrorException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import software.amazon.jdbc.HostListProvider;
+import software.amazon.jdbc.HostListProviderService;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.HostSpecBuilder;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.dialect.BlueGreenDialect;
+import software.amazon.jdbc.hostavailability.SimpleHostAvailabilityStrategy;
+import software.amazon.jdbc.hostlistprovider.RdsHostListProvider;
+import software.amazon.jdbc.plugin.iam.IamAuthConnectionPlugin;
+import software.amazon.jdbc.util.ConnectionUrlParser;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.PropertyUtils;
+import software.amazon.jdbc.util.RdsUtils;
+
+public class BlueGreenStatusMonitor {
+
+  private static final Logger LOGGER = Logger.getLogger(BlueGreenStatusMonitor.class.getName());
+  protected static final long DEFAULT_CHECK_INTERVAL_MS = TimeUnit.MINUTES.toMillis(5);
+  protected static final String BG_CLUSTER_ID = "941d00a8-8238-4f7d-bf59-771bff783a8e";
+
+  protected static final String latestKnownVersion = "1.0";
+
+  // Add more versions here if needed.
+  protected static final Set<String> knownVersions = new HashSet<>(Collections.singletonList(latestKnownVersion));
+  protected final BlueGreenDialect blueGreenDialect;
+  protected final PluginService pluginService;
+  protected final String bgdId;
+  protected final Properties props;
+  protected final BlueGreenRole role;
+  protected final OnStatusChange onStatusChangeFunc;
+
+  // Status check interval time in millis for each BlueGreenIntervalRate.
+  protected final Map<BlueGreenIntervalRate, Long> statusCheckIntervalMap;
+
+  protected final HostSpec initialHostSpec;
+
+  protected final ExecutorService executorService = Executors.newFixedThreadPool(1);
+  protected final ExecutorService openConnectionExecutorService = Executors.newFixedThreadPool(1);
```

Comment:
Can we use the ExecutorFactory classes to create threads with identifiable names